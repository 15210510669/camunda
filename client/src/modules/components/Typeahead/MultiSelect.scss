.MultiSelect {
  width: 100%;
  position: relative;

  .UncontrolledMultiValueInput {
    padding-right: 40px;

    .searchClear {
      right: 30px;
    }
  }

  .optionsButton {
    position: absolute;
    right: 0px;
    top: 0;
    height: 100%;
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
    background: none;
    border: none;
    max-height: 32px;
    padding: 0 12px 5px 0;

    .Icon {
      margin-right: 0px;
    }
  }
}
