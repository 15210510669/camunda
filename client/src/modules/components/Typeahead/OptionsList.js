/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useState, useEffect, useCallback} from 'react';
import {t} from 'translation';
import {Dropdown, LoadingIndicator} from 'components';

export default function OptionsList({
  loading,
  hasMore,
  input,
  onSelect,
  filter,
  children,
  onMouseDown,
  async,
  ...props
}) {
  const [selectedOption, setSelectedOption] = useState(-1);
  const optionList = React.createRef();
  const optionsArr = React.Children.toArray(children);
  let filteredOptions = optionsArr;

  if (!async) {
    filteredOptions = optionsArr.filter(({props: {label, children}}) =>
      (label || children).toLowerCase().includes(filter.toLowerCase())
    );
  }

  const optionsWithProps = filteredOptions.map((option, i) =>
    React.cloneElement(option, {
      className: i === selectedOption ? 'isActive' : '',
      onClick: evt => onSelect(option),
      onMouseDown
    })
  );

  const handleKeyPress = useCallback(
    evt => {
      const {open, onOpen, onClose} = props;
      let nextOption = -1;
      evt = evt || window.event;

      const optionsCount = optionsWithProps.length;

      if (evt.key === 'Enter') {
        evt.stopPropagation();
        const optionToSelect = optionsWithProps[selectedOption];
        if (optionToSelect) {
          onSelect(optionToSelect);
        }
        return;
      }

      if (evt.key === 'Escape') {
        return onClose();
      }

      if (evt.key === 'ArrowDown') {
        if (!open) {
          return onOpen();
        }
        nextOption = (selectedOption + 1) % optionsCount;
      }

      if (evt.key === 'ArrowUp') {
        nextOption = selectedOption - 1 < 0 ? optionsCount - 1 : selectedOption - 1;
      }

      if (optionList.current) {
        const selectedItem = optionList.current.querySelectorAll('.DropdownOption')[nextOption];
        if (selectedItem) {
          selectedItem.scrollIntoView({block: 'nearest', inline: 'nearest'});
        }
      }

      // scroll to end on the last element to show the has more info message
      if (nextOption === optionsCount - 1) {
        optionList.current.scrollTop = optionList.current.scrollHeight;
      }

      setSelectedOption(nextOption);
    },
    [selectedOption, onSelect, optionsWithProps, optionList, props]
  );

  useEffect(() => {
    if (!input) {
      return;
    }
    input.addEventListener('keydown', handleKeyPress);
    return () => {
      input.removeEventListener('keydown', handleKeyPress);
    };
  }, [handleKeyPress, input]);

  if (!props.open) {
    return null;
  }

  const isEmpty = filteredOptions.length === 0 && !loading;

  return (
    <div ref={optionList} className="OptionsList">
      {loading ? <LoadingIndicator /> : optionsWithProps}
      {isEmpty && <Dropdown.Option className="message">{t('common.notFound')}</Dropdown.Option>}
      {hasMore && (
        <Dropdown.Option className="message">
          {t('common.searchForMore', {count: optionsWithProps.length})}
        </Dropdown.Option>
      )}
    </div>
  );
}
