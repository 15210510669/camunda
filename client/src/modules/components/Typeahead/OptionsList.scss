.OptionsList {
  overflow-y: auto;
  overflow-x: hidden;
  border: 1px solid var(--silver-darken-80);
  border-radius: 3px;
  z-index: var(--z-popup);
  position: absolute;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
  background-color: var(--silver-lighten-99);
  width: 100%;
  max-height: 350px;

  &:empty {
    display: none;
  }

  .DropdownOption {
    white-space: nowrap;
    overflow-x: hidden;
    text-overflow: ellipsis;
    height: auto;
    &.isActive {
      background-color: rgba(216, 220, 227, 0.5);
      color: var(--grey-darken-13);
      border: none;
      transition: none;
    }

    &:hover {
      background-color: var(--blue-darken-55);
      color: var(--silver-lighten-99);
      .subText {
        color: var(--silver-lighten-99);
      }
    }

    &.message {
      color: var(--grey-lighten-50);
      padding-top: 10px;
      padding-bottom: 10px;

      &:hover,
      &:focus {
        color: var(--grey-lighten-50);
        background-color: var(--silver-lighten-99);
        cursor: auto;
      }
    }
  }
}
