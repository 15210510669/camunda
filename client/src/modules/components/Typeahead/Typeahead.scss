@import 'shared-styles/shared-styles.scss';

.Typeahead {
  width: 100%;
  position: relative;
  .Input {
    width: 100%;
    padding-left: 20px;

    &::placeholder {
      font-style: italic;
      color: #bbbbbb;
    }
  }

  .optionsButton {
    position: absolute;
    right: 0px;
    height: 100%;
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;

    .Icon {
      margin-right: 0px;
    }
  }

  .searchResultsList {
    overflow-y: auto;
    overflow-x: hidden;
    border: 1px solid #bbb;
    border-radius: 3px;
    z-index: 99999;
    position: absolute;
    box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
    background-color: white;

    .searchResult {
      &.isActive {
        background-color: rgba(216, 220, 227, 0.5);
        color: black;
        border: none;
        transition: none;
      }

      &:hover {
        background-color: $blue;
        color: #fff;
      }
    }
  }

  ::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: rgba(0, 0, 0, 0.5);
    -webkit-box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
  }
}
