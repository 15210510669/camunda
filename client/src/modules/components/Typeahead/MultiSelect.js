/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useRef, useState} from 'react';
import classnames from 'classnames';

import {Button, Icon, Dropdown, MultiValueInput} from 'components';
import {t} from 'translation';

import OptionsList from './OptionsList';
import Typeahead from './Typeahead';

import './MultiSelect.scss';

export default function MultiSelect({
  onSearch = () => {},
  onOpen = () => {},
  onClose = () => {},
  values = [],
  noValuesMessage,
  placeholder,
  className,
  onAdd,
  onRemove,
  async,
  onClear,
  typedOption,
  hasMore,
  loading,
  disabled,
  children,
}) {
  const [open, setOpen] = useState(false);
  const [query, setQuery] = useState('');
  const [optionClicked, setOptionClicked] = useState(false);
  const [resetInput, setResetInput] = useState(false);
  const [insideClick, setInsideClick] = useState(false);

  const input = useRef();

  const isEmpty = !loading && !query && React.Children.count(children) === 0;
  const isInputDisabled = disabled || (isEmpty && !typedOption);

  function showList() {
    if (!open) {
      onOpen(query);
      setOpen(true);
    }
  }

  function hideList() {
    if (!optionClicked) {
      if (query && !insideClick) {
        setResetInput(!resetInput);
        setQuery('');
      }
      onClose();
      setOpen(false);
    } else {
      input.current.focus();
    }
    setInsideClick(false);
  }

  function selectOption({props: {value}}) {
    onAdd(value);
    setOptionClicked(false);
  }

  function getPlaceholderText() {
    if (disabled || isEmpty) {
      return noValuesMessage || t('common.notFound');
    }
    return placeholder;
  }

  function onChange(value) {
    setQuery(value);
    onSearch(value);
  }

  return (
    <div className={classnames('MultiSelect', className)} onMouseDown={() => setInsideClick(true)}>
      <MultiValueInput
        className="typeaheadInput"
        key={resetInput}
        onChange={onChange}
        values={values}
        onAdd={(...args) => {
          if (!optionClicked) {
            onAdd(...args);
          }
        }}
        onRemove={onRemove}
        ref={input}
        onBlur={hideList}
        onClear={onClear}
        onFocus={showList}
        disabled={isInputDisabled}
        placeholder={getPlaceholderText(isEmpty)}
        disableAddByKeyboard
      />
      <Button
        tabIndex="-1"
        className="optionsButton"
        onClick={() => input.current.focus()}
        disabled={isInputDisabled}
      >
        <Icon type="down" className="downIcon" />
      </Button>
      <OptionsList
        async={async}
        open={open}
        onClose={hideList}
        onOpen={open}
        filter={query}
        onSelect={selectOption}
        input={input.current}
        onMouseDown={() => setOptionClicked(true)}
        loading={loading}
        hasMore={hasMore}
        typedOption={typedOption}
      >
        {children}
      </OptionsList>
    </div>
  );
}

MultiSelect.Option = Dropdown.Option;

MultiSelect.Highlight = Typeahead.Highlight;
