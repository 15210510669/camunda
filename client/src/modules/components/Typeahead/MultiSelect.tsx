/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {ReactElement, useEffect, useRef, useState, Children, ReactNode} from 'react';
import classnames from 'classnames';

import {Button, Icon, Dropdown, UncontrolledMultiValueInput} from 'components';
import {t} from 'translation';

import OptionsList, {OptionProps} from './OptionsList';
import Typeahead from './Typeahead';

import './MultiSelect.scss';

type MultiSelectProps = {
  onSearch?: (query: string) => void;
  onOpen?: (query: string) => void;
  onClose?: () => void;
  values?: {label: string; value: string}[];
  noValuesMessage?: string;
  placeholder?: ReactNode;
  className?: string;
  onAdd: (value: string) => void;
  onRemove: (value: string, index: number) => void;
  async?: boolean;
  onClear: () => void;
  typedOption?: boolean;
  hasMore?: boolean;
  loading?: boolean;
  disabled?: boolean;
  children: ReactNode;
  persistMenu?: boolean;
};

export default function MultiSelect({
  onSearch = () => {},
  onOpen = () => {},
  onClose = () => {},
  values = [],
  noValuesMessage,
  placeholder,
  className,
  onAdd,
  onRemove,
  async,
  onClear,
  typedOption,
  hasMore,
  loading,
  disabled,
  children,
  persistMenu = true,
}: MultiSelectProps): JSX.Element {
  const [open, setOpen] = useState(false);
  const [query, setQuery] = useState('');
  const [insideClick, setInsideClick] = useState(false);

  const input = useRef<HTMLInputElement>(null);

  const isEmpty = !loading && !query && Children.count(children) === 0;
  const isInputDisabled = disabled || (isEmpty && !typedOption);

  useEffect(() => {
    if (isInputDisabled) {
      setOpen(false);
    }
  }, [isInputDisabled]);

  function showList() {
    if (!open) {
      onOpen(query);
      setOpen(true);
    }
  }

  function hideList() {
    if (!insideClick) {
      onClose();
      setOpen(false);
    } else {
      // wait for input to blur before focusing again
      setTimeout(() => input.current?.focus(), 0);
    }
    setInsideClick(false);
  }

  function selectOption({props: {value}}: ReactElement<OptionProps>) {
    if (query) {
      setQuery('');
      onSearch('');
    }
    if (!persistMenu) {
      hideList();
    }
    onAdd(value);
  }

  function getPlaceholderText() {
    if (disabled || isEmpty) {
      return noValuesMessage || t('common.notFound');
    }
    return placeholder;
  }

  function onChange({target: {value}}: React.ChangeEvent<HTMLInputElement>) {
    setQuery(value);
    onSearch(value);
    showList();
  }

  function handleKeyPress(evt: React.KeyboardEvent<HTMLInputElement>) {
    if (query === '' && evt.key === 'Backspace' && values.length > 0) {
      const lastElementIndex = values.length - 1;
      const lastElementValue = values[lastElementIndex]?.value;
      if (lastElementValue) {
        onRemove(lastElementValue, lastElementIndex);
      }
    }
  }

  return (
    <div
      className={classnames('MultiSelect', className)}
      onMouseDown={() => setInsideClick(true)}
      onMouseUp={() => setInsideClick(false)}
    >
      <UncontrolledMultiValueInput
        inputClassName="typeaheadInput"
        ref={input}
        value={query}
        onChange={onChange}
        values={values}
        onRemove={onRemove}
        onBlur={hideList}
        onClear={onClear}
        onFocus={showList}
        disabled={isInputDisabled}
        placeholder={getPlaceholderText()}
        onKeyDown={handleKeyPress}
      />
      <Button
        tabIndex={-1}
        className="optionsButton"
        onClick={() => input.current?.focus()}
        disabled={isInputDisabled}
      >
        <Icon type="down" className="downIcon" />
      </Button>
      <OptionsList
        async={async}
        open={open}
        onClose={hideList}
        onOpen={showList}
        filter={query}
        onSelect={selectOption}
        input={input.current}
        loading={loading}
        hasMore={hasMore}
        typedOption={typedOption}
      >
        {children}
      </OptionsList>
    </div>
  );
}

MultiSelect.Option = Dropdown.Option;

MultiSelect.Highlight = Typeahead.Highlight;
