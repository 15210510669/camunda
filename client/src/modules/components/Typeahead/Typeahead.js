/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import classnames from 'classnames';
import {Input, Dropdown} from 'components';
import {formatters} from 'services';

import './Typeahead.scss';

const valuesShownInBox = 10;
const valueHeight = 30;

export default class Typeahead extends React.Component {
  state = {
    query: '',
    optionsVisible: false,
    selectedValueIdx: 0,
    firstShownOptionIdx: 0,
    lastCommittedValue: ''
  };

  componentDidMount() {
    document.body.addEventListener('click', this.close);
  }

  componentDidUpdate(prevProps) {
    if (this.props.initialValue && this.props.initialValue !== prevProps.initialValue) {
      const value = this.getFormatter()(this.props.initialValue);
      this.setState({
        query: value,
        lastCommittedValue: value
      });
    }
    if (this.props.disabled !== prevProps.disabled) {
      this.setState({
        query: this.props.disabled ? this.props.placeholder : '',
        lastCommittedValue: ''
      });
    }
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.close);
  }

  returnFocusToInput = evt => {
    if (evt.target === this.optionsList) {
      this.input.focus();
    }
  };

  showOptions = evt => {
    if (evt && evt.type === 'click') {
      this.input.select();
    }

    if (!this.state.optionsVisible) {
      this.setState({
        optionsVisible: true
      });
    }
  };

  close = evt => {
    if (!evt || !this.container.contains(evt.target)) {
      this.setState({
        optionsVisible: false
      });
    }
  };

  updateValues = () => {
    this.setState({
      selectedValueIdx: 0,
      firstShownOptionIdx: 0
    });
  };

  updateQuery = async evt => {
    this.setState({query: evt.target.value});
    this.showOptions();
    this.updateValues();
  };

  getFormatter = () => this.props.formatter || (v => v);

  selectValue = value => () => {
    const formatter = this.getFormatter();
    const formattedValue = formatter(value);
    this.setState({
      query: formattedValue,
      lastCommittedValue: formattedValue,
      selectedValueIdx: 0,
      firstShownOptionIdx: 0,
      optionsVisible: false
    });

    if (this.props.onSelect) {
      this.props.onSelect(value);
    }
    this.close();
  };

  resetToLastCommitted = () => {
    this.setState({query: this.state.lastCommittedValue});
  };

  handleKeyPress = evt => {
    evt = evt || window.event;

    if (evt.key === 'Tab') {
      this.close();
      return;
    }

    const {selectedValueIdx, optionsVisible} = this.state;
    const values = this.getFilteredValues();
    if (evt.key === 'Enter') {
      evt.preventDefault();
      if (optionsVisible && values[selectedValueIdx]) {
        this.selectValue(values[selectedValueIdx])();
        evt.stopPropagation();
      }
      return;
    }

    if (evt.key === 'Escape') {
      if (this.state.optionsVisible) {
        evt.stopPropagation();
      }
      this.close();
    } else {
      let selectedValueIdx = this.state.selectedValueIdx;

      if (evt.key === 'ArrowDown') {
        evt.preventDefault();
        if (!this.state.optionsVisible) {
          this.showOptions();
        } else {
          selectedValueIdx = (selectedValueIdx + 1) % values.length;
        }
      }

      if (evt.key === 'ArrowUp') {
        evt.preventDefault();
        selectedValueIdx = selectedValueIdx - 1 < 0 ? values.length - 1 : selectedValueIdx - 1;
      }

      const firstShownOptionIdx = this.scrollIntoView(selectedValueIdx);

      this.setState({
        selectedValueIdx,
        firstShownOptionIdx
      });
    }
  };

  scrollIntoView = selectedValueIdx => {
    let {firstShownOptionIdx} = this.state;
    const values = this.getFilteredValues();

    if (this.optionsList) {
      if (selectedValueIdx === 0) {
        this.optionsList.scrollTop = 0;
        firstShownOptionIdx = 0;
      }

      if (selectedValueIdx === values.length - 1) {
        this.optionsList.scrollTop = this.optionsList.scrollHeight;
        firstShownOptionIdx = values.length - valuesShownInBox;
      }

      if (selectedValueIdx >= firstShownOptionIdx + valuesShownInBox) {
        this.optionsList.scrollTop = (firstShownOptionIdx + 1) * valueHeight;
        firstShownOptionIdx++;
      }

      if (selectedValueIdx < firstShownOptionIdx) {
        firstShownOptionIdx--;
        this.optionsList.scrollTop = selectedValueIdx * valueHeight;
      }
    }

    return firstShownOptionIdx;
  };

  optionsListRef = optionsList => {
    this.optionsList = optionsList;
    if (optionsList) {
      this.optionsList.highestWidth = this.optionsList.offsetWidth;
    }
  };

  containerRef = container => {
    this.container = container;
  };

  inputRef = input => {
    this.input = input;
  };

  getFilteredValues = () =>
    this.props.values.filter(value =>
      this.getFormatter()(value)
        .toLowerCase()
        .includes(this.state.query.toLowerCase())
    );

  render() {
    const formatter = this.getFormatter();

    const {query, selectedValueIdx, optionsVisible} = this.state;
    const values = this.getFilteredValues();

    const searchResultContainerStyle = {
      maxHeight: valueHeight * valuesShownInBox + 'px',
      maxWidth: this.optionsList && this.optionsList.highestWidth + 'px'
    };

    const valueStyle = {
      height: valueHeight + 'px',
      minWidth: this.input && this.input.clientWidth + 'px'
    };

    const hasNoValues = this.props.values.length === 0;

    return (
      <div ref={this.containerRef} className={classnames('Typeahead', this.props.className)}>
        <Input
          className={classnames({isInvalid: this.props.isInvalid})}
          value={hasNoValues ? 'No results found' : query}
          onChange={this.updateQuery}
          onClick={this.showOptions}
          onFocus={this.showOptions}
          onKeyDown={this.handleKeyPress}
          onBlur={this.resetToLastCommitted}
          placeholder={this.props.placeholder}
          ref={this.inputRef}
          disabled={this.props.disabled || hasNoValues}
        />
        {optionsVisible && values.length > 0 && (
          <div
            style={searchResultContainerStyle}
            className="searchResultsList"
            ref={this.optionsListRef}
            onMouseUp={this.returnFocusToInput}
          >
            {values.map((value, index) => {
              return (
                <Dropdown.Option
                  className={classnames('searchResult', {
                    isActive: index === selectedValueIdx
                  })}
                  style={valueStyle}
                  onMouseDown={evt => evt.preventDefault()}
                  onClick={this.selectValue(value)}
                  key={value && value.id ? value.id : formatter(value)}
                >
                  {formatters.getHighlightedText(formatter(value), this.state.query)}
                </Dropdown.Option>
              );
            })}
          </div>
        )}
      </div>
    );
  }
}
