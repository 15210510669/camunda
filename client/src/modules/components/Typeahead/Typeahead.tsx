/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {
  ReactElement,
  createRef,
  ChangeEvent,
  Component,
  Children,
  isValidElement,
  ReactNode,
} from 'react';

import {Input, Button, Icon, Dropdown} from 'components';
import {t} from 'translation';
import classnames from 'classnames';

import OptionsList from './OptionsList';

import './Typeahead.scss';

interface TypeaheadProps {
  children: ReactNode;
  initialValue?: string;
  value?: string;
  onSearch: (query: string) => void;
  onChange: (value: string) => void;
  onOpen: () => void;
  onClose: () => void;
  placeholder?: string | JSX.Element[];
  disabled?: boolean;
  loading?: boolean;
  hasMore?: boolean;
  async?: boolean;
  typedOption?: boolean;
  noValuesMessage?: string | JSX.Element[];
  className?: string;
}

interface TypeaheadState {
  query: string;
  open: boolean;
  selected: string | undefined;
}

const defaultState: TypeaheadState = {
  query: '',
  open: false,
  selected: '',
};

export default class Typeahead extends Component<TypeaheadProps, TypeaheadState> {
  static defaultProps = {
    onSearch: () => {},
    onChange: () => {},
    onOpen: () => {},
    onClose: () => {},
  };

  state = defaultState;

  optionClicked = false;

  input = createRef<HTMLInputElement>();

  onTextChange = (evt: ChangeEvent<HTMLInputElement>) => {
    this.setState({query: evt.target.value, open: true});
    this.props.onSearch(evt.target.value);
  };

  componentDidMount() {
    const {initialValue, value} = this.props;

    this.findAndSelect(initialValue);
    this.findAndSelect(value);
  }

  componentDidUpdate(prevProps: TypeaheadProps) {
    if (prevProps.value !== this.props.value) {
      if (typeof this.props.value === 'undefined') {
        this.setState(defaultState);
        this.props.onSearch('');
      } else {
        this.findAndSelect(this.props.value);
      }
    }
    if (!prevProps.initialValue && this.props.initialValue) {
      this.findAndSelect(this.props.initialValue);
    }
  }

  findAndSelect = (value?: string) => {
    const {children, typedOption} = this.props;
    const foundOption = Children.toArray(children).find(
      (option) => (option as ReactElement).props.value === value
    );

    if (typedOption) {
      this.setState({
        selected: value,
        query: value || '',
      });
    } else if (isValidElement(foundOption)) {
      const {label, children} = foundOption.props;
      const selected = label || children;
      this.setState({
        selected,
        query: selected,
      });
    }
  };

  selectOption = ({props: {label, children, value}}: ReactElement) => {
    const selected = label || children;
    this.setState({
      selected,
      query: selected,
      open: false,
    });
    this.props.onChange(value);
    this.optionClicked = false;
  };

  getPlaceholderText = (isEmpty: boolean) => {
    const {placeholder, disabled, noValuesMessage} = this.props;
    const {selected} = this.state;
    if (selected) {
      return selected;
    }
    if (disabled || isEmpty) {
      return noValuesMessage || t('common.notFound');
    }
    return placeholder;
  };

  open = () => {
    this.props.onOpen();
    this.setState({open: true, query: ''});
  };

  close = () => {
    this.props.onClose();
    this.setState(({selected}) => ({open: false, query: selected || ''}));
  };

  render() {
    const {children, disabled, loading, hasMore, async, typedOption, className} = this.props;
    const {query, open, selected} = this.state;
    const isEmpty = !loading && !query && !typedOption && Children.count(children) === 0;
    const isInputDisabled = isEmpty || disabled;

    return (
      <div className={classnames(className, 'Typeahead')}>
        <Input
          type="text"
          className={classnames('typeaheadInput', {selectionVisible: open && selected})}
          value={query}
          onFocus={this.open}
          onBlur={(evt) => {
            if (!this.optionClicked) {
              this.close();
            }
          }}
          onChange={this.onTextChange}
          ref={this.input}
          placeholder={this.getPlaceholderText(isEmpty)}
          disabled={isInputDisabled}
        />
        <Button
          tabIndex={-1}
          className="optionsButton"
          onClick={() => this.input.current?.focus()}
          disabled={isInputDisabled}
        >
          <Icon type="down" className="downIcon" />
        </Button>
        <OptionsList
          async={async}
          open={open}
          onClose={this.close}
          onOpen={this.open}
          filter={query}
          onSelect={this.selectOption}
          input={this.input.current}
          onMouseDown={() => (this.optionClicked = true)}
          loading={loading}
          hasMore={hasMore}
          typedOption={typedOption}
        >
          {children}
        </OptionsList>
      </div>
    );
  }

  static Option = Dropdown.Option;

  static Highlight = function Highlight(props: {children: string; matchFromStart?: boolean}) {
    return <>{props.children}</>;
  };
}
