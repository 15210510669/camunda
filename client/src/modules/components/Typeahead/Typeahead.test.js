import React from 'react';

import Typeahead from './Typeahead';
import {Dropdown, Input} from 'components';

import {mount} from 'enzyme';

it('should display available options if dropdown is open', () => {
  const node = mount(<Typeahead values={['foo', 'bar']} />);

  node.setState({
    optionsVisible: true
  });

  expect(node.find(Dropdown.Option).at(0)).toIncludeText('foo');
  expect(node.find(Dropdown.Option).at(1)).toIncludeText('bar');
});

it('should format the data based on the provided formatter', () => {
  const node = mount(<Typeahead values={['foo', 'bar']} formatter={v => v + v} />);

  node.setState({
    optionsVisible: true
  });

  expect(node.find(Dropdown.Option).at(0)).toIncludeText('foofoo');
  expect(node.find(Dropdown.Option).at(1)).toIncludeText('barbar');
});

it('should only display entries that match the typeahead value', () => {
  const node = mount(<Typeahead values={['varFoo', 'varBar', 'varFoobar']} />);

  node.setState({
    optionsVisible: true,
    query: 'foo'
  });

  expect(node.find(Dropdown.Option).at(0)).toIncludeText('varFoo');
  expect(node.find(Dropdown.Option).at(1)).toIncludeText('varFoobar');
  expect(node).not.toIncludeText('varBar');
});

it('should only display entries that match the typeahead value, even if there is a formatter', () => {
  const node = mount(
    <Typeahead values={['varFoo', 'varBar', 'varFoobar']} formatter={v => v + v} />
  );

  node.setState({
    optionsVisible: true,
    query: 'foobarvar'
  });

  expect(node.find(Dropdown.Option).at(0)).toIncludeText('varFoobarvarFoobar');
});

it('should call the provided onSelect method when a selection is done', () => {
  const spy = jest.fn();
  const node = mount(<Typeahead values={['foo', 'bar']} onSelect={spy} />);

  node.setState({
    optionsVisible: true
  });

  node
    .find(Dropdown.Option)
    .at(0)
    .simulate('click');

  expect(spy).toHaveBeenCalledWith('foo');
});

it('should reset the query to the latest committed state when the input field blurs', () => {
  const node = mount(<Typeahead values={['foo', 'bar']} />);

  node.setState({
    optionsVisible: true,
    lastCommittedValue: 'foo',
    query: 'I typed something in the typeahead'
  });

  node.find(Input).simulate('blur');

  expect(node).toHaveState('query', 'foo');
});

it('should disable the input field when disabled prop is set to true', () => {
  const node = mount(<Typeahead values={['foo', 'bar']} disabled={true} />);

  expect(node.find(Input)).toHaveProp('disabled', true);
});

it('should show the dropdown menu options on input focus', () => {
  const node = mount(<Typeahead values={['foo', 'bar']} />);
  node.find(Input).prop('onFocus')();

  expect(node).toHaveState('optionsVisible', true);
});
