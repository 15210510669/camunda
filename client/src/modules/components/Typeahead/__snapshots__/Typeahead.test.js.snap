// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display available options if dropdown is open 1`] = `
<div
  className="Typeahead"
>
  <ForwardRef(Input)
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    value=""
  />
  <ForwardRef(Button)
    className="optionsButton"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="downIcon"
      type="down"
    />
  </ForwardRef(Button)>
  <div
    className="searchResultsList"
    onMouseUp={[Function]}
    style={
      Object {
        "maxHeight": "300px",
        "maxWidth": undefined,
      }
    }
  >
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="foo"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      foo
    </ForwardRef(DropdownOption)>
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="bar"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      bar
    </ForwardRef(DropdownOption)>
  </div>
</div>
`;

exports[`should format the data based on the provided formatter 1`] = `
<div
  className="Typeahead"
>
  <ForwardRef(Input)
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    value=""
  />
  <ForwardRef(Button)
    className="optionsButton"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="downIcon"
      type="down"
    />
  </ForwardRef(Button)>
  <div
    className="searchResultsList"
    onMouseUp={[Function]}
    style={
      Object {
        "maxHeight": "300px",
        "maxWidth": undefined,
      }
    }
  >
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="foofoo"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      foofoo
    </ForwardRef(DropdownOption)>
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="barbar"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      barbar
    </ForwardRef(DropdownOption)>
  </div>
</div>
`;

exports[`should only display entries that match the typeahead value 1`] = `
<div
  className="Typeahead"
>
  <ForwardRef(Input)
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    value="foo"
  />
  <ForwardRef(Button)
    className="optionsButton"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="downIcon"
      type="down"
    />
  </ForwardRef(Button)>
  <div
    className="searchResultsList"
    onMouseUp={[Function]}
    style={
      Object {
        "maxHeight": "300px",
        "maxWidth": undefined,
      }
    }
  >
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="varFoo"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      <span
        key="0"
      >
        var
      </span>
      <span
        className="textBold"
        key="1"
      >
        Foo
      </span>
      <span
        key="2"
      />
    </ForwardRef(DropdownOption)>
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="varFoobar"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      <span
        key="0"
      >
        var
      </span>
      <span
        className="textBold"
        key="1"
      >
        Foo
      </span>
      <span
        key="2"
      >
        bar
      </span>
    </ForwardRef(DropdownOption)>
  </div>
</div>
`;

exports[`should only display entries that match the typeahead value, even if there is a formatter 1`] = `
<div
  className="Typeahead"
>
  <ForwardRef(Input)
    className=""
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    value="foobarvar"
  />
  <ForwardRef(Button)
    className="optionsButton"
    disabled={false}
    onClick={[Function]}
  >
    <Icon
      className="downIcon"
      type="down"
    />
  </ForwardRef(Button)>
  <div
    className="searchResultsList"
    onMouseUp={[Function]}
    style={
      Object {
        "maxHeight": "300px",
        "maxWidth": undefined,
      }
    }
  >
    <ForwardRef(DropdownOption)
      className="searchResult"
      key="varFoobarvarFoobar"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "height": "30px",
          "minWidth": undefined,
        }
      }
    >
      <span
        key="0"
      >
        var
      </span>
      <span
        className="textBold"
        key="1"
      >
        Foobarvar
      </span>
      <span
        key="2"
      >
        Foobar
      </span>
    </ForwardRef(DropdownOption)>
  </div>
</div>
`;
